src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+        {
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, relativeUri),
+                Method = HttpMethod.Get
+            };
+
+            return await GetHttpContent(credentialsRequest, "Failed to acquire ECS credentials.").ConfigureAwait(false);
+        }
+
+        private static async Task<string> GetHttpContent(HttpRequestMessage request, string exceptionMessage)
+        {
+            HttpResponseMessage response;
+            try
+            {
+                response = await _httpClientInstance.Value.SendAsync(request).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            catch (HttpRequestException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new MongoClientException(exceptionMessage);
+            }
+
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        }
+
+        private static void ValidateCredentials(string username, string password, string sessionToken)
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        private static void ValidateCredentials(string username, SecureString password, string sessionToken)
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        // nested classes
+        private class AwsCredentials
+        {
+            public UsernamePasswordCredential Credentials;
+            public string SessionToken;
+        }
+
+        private class MongoAWSMechanism : ISaslMechanism
+        {
+            private readonly UsernamePasswordCredential _credential;
+            private readonly string _name;
+            private readonly IRandomByteGenerator _randomByteGenerator;
+            private readonly string _sessionToken;
+
+            public MongoAWSMechanism(
+                UsernamePasswordCredential credential,
+                string sessionToken,
+                IRandomByteGenerator randomByteGenerator)
+            {
+                _name = "MONGODB-AWS";
```

Comment:
it's better to take this value from the main class. See: https://github.com/DmitryLukyanov/mongo-csharp-driver/commit/1b03be95b0bd47b12600bc634bcbe112da1d866f#diff-ffb35b98eecd05d3c7b6133327bf13b3R242